generator/json-generator.lua

```diff
@@ -1,135 +1,214 @@
         end
     end
+end
 
-    -- Generate the snippets for all LÖVE modules.
-    print( TAB .. 'Writing modules' )
-    for i, module in ipairs( api.modules ) do
-        print( TAB .. TAB .. '- ' .. module.name )
-
-        if module.types then
-            for _, type in ipairs( module.types ) do
-                print( TAB .. TAB .. TAB .. '-> ' .. type.name )
-
-                for _, f in ipairs( type.functions ) do
-                    local str = buildTypeFunctionCompletion( f, type )
-                    for l = 1, #str do
-                        file:write( str[l] )
-                    end
-                end
+local function createGlobals( api, global )
+    global.type = 'table'
+    global.fields = {}
+    global.fields.love = {}
+    global.fields.love.type = 'table'
+    global.fields.love.fields = {}
+
+    createLoveFunctions( api.functions, global.fields.love.fields )
+    createLoveFunctions( api.callbacks, global.fields.love.fields )
+    createLoveModuleFunctions( api.modules, global.fields.love.fields )
+end
+
+-- ------------------------------------------------
+-- Creation methods for named types
+-- ------------------------------------------------
+
+
+local function collectTypes( api )
+    local types = {}
+
+    for _, type in pairs( api.types ) do
+        assert( not types[type.name] )
+        types[type.name] = type
+    end
+
+    for _, module in pairs( api.modules ) do
+        if module.types then -- Not all modules define types.
+            for _, type in pairs( module.types ) do
+                assert( not types[type.name] )
+                types[type.name] = type
             end
         end
+    end
+
+    return types
+end
+
+local function createTypes( types, namedTypes )
+    namedTypes.Types = {}
+    namedTypes.Types.type = 'table'
+    namedTypes.Types.fields = {}
+
+    for name, type in pairs( types ) do
+        print( '  ' .. name )
+        namedTypes.Types.fields[name] = {}
+        namedTypes.Types.fields[name].type = 'table'
+        namedTypes.Types.fields[name].fields = {}
+
+        namedTypes.Types.fields[name].metatable = {}
+        namedTypes.Types.fields[name].metatable.type = 'table'
 
-        for j, f in ipairs( module.functions ) do
-            local str = buildModuleFunctionCompletion( f, module, ( i == #api.modules and j == #module.functions )  )
-            for k = 1, #str do
-                file:write( str[k] )
+        namedTypes.Types.fields[name].metatable.fields = {}
+
+        namedTypes.Types.fields[name].metatable.fields.__index = {}
+        namedTypes.Types.fields[name].metatable.fields.__index.type = 'table'
+        namedTypes.Types.fields[name].metatable.fields.__index.fields = {}
+
+        if type.functions then
+            for _, f in pairs( type.functions ) do
+                createFunction( f, namedTypes.Types.fields[name].metatable.fields.__index.fields, name .. ':' )
```

Comment:
I would use the table part for the methods that belong to the Type and leave the `__index` clear for now.

Do you get completions with that?